### ISSUE TITLE
CDI with EJB in Payara/Glassfish 4 is still an issue for EAR

### ISSUE THREAD

**Issue by rsoika** [payara/Payara#2508]
# Description #
----------
I think there is still a deployment issue with glassfish when it comes to CDI and EJB in a EAR file.

```
myapp.ear/
  | - some-lib-with ejb-and-cdi.jar
  | - my-ejb.jar
  | - my-web.war
  | - lib/
       | - some-other-lisb-without ejb/cdi.jar
```

CDI Beans contained in a JAR module with EJBs result in a deployment where the EJB can not see the CDI beans. If you deploy all the stuff in a single WAR file everything works. 

```
myapp.war/
  | - lib/
      | - some-lib-with ejb-and-cdi.jar
      | - my-ejb.jar
      | - some-other-lisb-without ejb/cdi.jar
```

If you ask why CDI is needed in  EJB a typical example would be the CDI Observer pattern which should work since Java EE6

- http://www.adam-bien.com/roller/abien/entry/java_ee_6_observer_with

Another discussion can be found here:

- https://stackoverflow.com/questions/3175458/cdi-with-ejb-3-1-weld-logger-on-glassfish-v3-0-1
- https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/payara-forum/oaKfCRSJLyw/Ax-mi2kuCAAJ

## Expected Outcome
The expected outcome is that a ejb jar file contained in an EAR shows the same behavior as if the ejb jar is deployed in a WAR.

## Current Outcome

EJBs Jar files with CDI is not working when deployed to the EAR root. The CDI beans result to null.
Only when deploying all as an .war it works as expected 

## Environment ##

I tested this behavior with payara/server-full:181
JDK 8
Linux/Docker

**Comment by smillidge** [comment]
Can you create a simple test case application on GitHub that reproduces the issue?

**Comment by lprimak** [comment]
I can confirm this behavior.  It bit me in the past.  This is an upstream bug / missing feature I believe

**Comment by smillidge** [comment]
@lprimak do you have a test case then?

**Comment by lprimak** [comment]
Sorry no, but I have seen this in the past

**Comment by rsoika** [comment]
I have created a test project here:

https://github.com/rsoika/payara-issue-2508

**Comment by smillidge** [comment]
Thanks for the test case we will try to reproduce. The behaviour of CDI in EAR files is poorly defined.

**Comment by rsoika** [comment]
This issue helped me to think about why I did not switch to WAR modules in general. It makes not really sense to build EARs in a time when we all talk more and more about microservices and container based architectures.
From this perspective, maybe EARs somehow have become unnecessary. Packaging all into a war makes it much easier and my application works as expected. So do not investigate to much of your time.  Payara is a great project!

**Comment by lprimak** [comment]
There is still a good place for EAR files, like when some libraries are reused across modules, like entities and business logic classes

**Comment by rsoika** [comment]
Yes, that's true and it is exactly my case. But on the other hand, in my case the libraries take about 500 KB of the entire application. Is this worth to relinquish on interoperability? I agree with you that EAR classloading was always a tricky thing. 
Would not it be better, if the next Java EE introduces a new clear concept of 'shared libraries' which can be easily integrated in a EJB, WAR or EAR module? If you want to run multible instances of the same application in one Application Server there is today no concept how to describe a dependency to a shared library - like it is possible e.g. in maven. But I am no server expert, and so you have a better overview about all these things.

**Comment by lprimak** [comment]
I was able to reproduce this issue with:
- deploy the application .ear file
- ```curl -D - http://localhost:8080/test/rest-service/test```

**Comment by lprimak** [comment]
Created internal issue ```PAYARA-2695```

**Comment by lprimak** [comment]
PR https://github.com/rsoika/payara-issue-2508/pull/1

**Comment by lprimak** [comment]
I looked more closely at your project, and I think you have a misunderstanding how library resolution works in EAR files.
It is different, but also valid from the WAR files.
You can't just add a classpath manifest entry into ejb-jar, and put it in the root of an EAR and expect it to work.
Ear lib/ (s) are shared libraries, which can contain CDI, Java classes, interfaces, etc. but not EJBs.
EJB modules can reference CDI in shared libraries

**Comment by lprimak** [comment]
If this satisfies, please close this issue

**Comment by rsoika** [comment]
I know what you mean and I looked into your changes of my example project. And I think I am absolute clear about what you are writing. 
But the situation is a little bit more tricky. I can't move the EJB class into the EJB module. 

The reason is (and my example is maybe misunderstanding) my 'lib' contains shared EJBs! So this is a kind of shared EJB-Library.
To understand this you may take a short look into our project [Imixs-Workflow](https://github.com/imixs/imixs-workflow). 
What we are doing here is the following:

- We  provide a library containing full featured EJBs (including timer services, interceptors and so on). You can find this sources[here](https://github.com/imixs/imixs-workflow/tree/master/imixs-workflow-engine/src/main).
- To include and activate the EJBs of the Imixs-Workflow Engine in a custom EAR project it is necessary to provide a custom EJB module with at least an empty ejb-jar.xml.
-  To import the EJBs form the Imxis-Workflow project, a Manifest classpath entry is needed. 

The reason for that setup is, that a custom project may customize the behavior of the Imixs EJB provided by the Imixs-Workflow engine. For example it is possible to customize the interceptors or provide custom resources for Role-Mapping or Mail Service injection. For that reason we are not including a ejb-jar.xml file into the Imixs-Workflow engine.

Why we are doing this? Our goal is to provide a shared ejb-library with an Java EE Workflow Engine. For my understanding this was the initial idea of EJBs in the very beginning of Java Enterprise. But most enterprise projects do not use this feature. I discussed this in the beginning of Java EE5 with some guys from Sun, and they advised me to this setup. We are running projects in enterprises on Glassfish2. And this setup works also for Wildfly very well.  But since Glassfish3 it seems to me, that this idea of an complete customize able shared EJB library got lost?

When I deploy the EJB-Library into a WAR file or a microservice there is absolutely no problem. Neither in Glassfish nor in Wildfly. The deployment process recognizes all EJBs provided in libs and with a ejb-jar.xml file I can customize the behavior of the external EJBs.

It looks like I am the only one who is providing an open source project with this feature.  I hope that with Jakarta EE 9 and the next stage of harmonizing EJBs and CDI we got rid of the deployment problem.I hope you can follow my thoughts and don't think I'm completely crazy.

**Comment by rsoika** [comment]
To be a little bit more precise, Payara is recognizing share-ejb-libs in an EAR correctly. And the external EJBs are deployed and working well in Payara. The only problem occurs when these EJBs are using CDI events. I think this is really an issue of combination of EJB with CDI. But Java EE is developing in this direction. So this is why I think it is an issue at the end.

**Comment by lprimak** [comment]
I've taken a look at your project, and I think what I have done in my pull request is valid, can solve your problem, and is correct according to Java EE specification.
What you are trying to do is not correct according to Java EE specification.

You should split up your services into API (interfaces, annotations, etc) and the EJB implementation (or CDI implementation) themselves.
EJB/CDI can be an ejb-jar inside the EAR file, and the API can be shared in EAR /lib directory.

Am I understanding your intent correctly or is there something else that I am missing?
I understand that it may be a PITA to split up code into API/implementation but as it stands right now, it's the only way to make this into a compliant EJB application

**Comment by rsoika** [comment]
I think that I understand you. But this is not what I try to achieve. 
Now you have again put the EJB Implementation together with the ejb-jar.xml in one jar. But if the jar is an external library form another project, than you either can not change the ejb-jar.xml, or you do not have the sources. 
 
I tagged your version (0.0.2) and tried to extend the example with an additional impl.jar, just to split interfaces from implementation. But I failed to deploy. 

Why do I use this manifest Class-Path entries: I want to extend the classpath of a custom EJB module with classes provided by an external EJB Impl LIb. I do not think that not splitting interfaces from EJB-Impl is the problem here. 

Compare this all to a deployment setup in a WAR file. In this case you just put all the libs (EJBs, CDIs, Pojos, EntityBeans) into the WEB-INF/lib/
And than you put a ejb-jar.xml into the /WEB-INF/ to declare you EJB setup. beans.xml, glassfish-ejb-jar.xml and glassfish-web.xml are also placed into /WEB-INF/. And everything works.
(the only curiosity is that you have to place the persistenc.xml into /classes/META-INF). But this is not our point here ;-)

The deployer for the .war artecfact can handle this.
But if you try to package all this stuff in .ear the ear deployer seems to fail. 

I fear that my English is not good enough to explain this idea of an external EJB-Lib without deployment descriptors.

But maybe you can take now a look at my latest setup of the example project. I added an additional ejb-impl module just to preserve your split of Interfaces and Impl code. But where should I put this lib with the Impl code when this is a closed jar file (imaging you do not have the sources because the code comes form an external project).

**Comment by rsoika** [comment]
I finally narrowed to a solution. After I declared the external Lib as an EJB module in the application.xml it works!. There is no need to split interfaces and impl. THere is no need to add custom manifest classpath entries. 
Take a look at my final [example](https://github.com/rsoika/payara-issue-2508).

**But:** If I now try to overwrite the behavior of this external ejb-lib by declaring the EJB in the ejb-jar.xml of the custom EJB module the deployment failed. You can test it with the empty ejb-jar.xml file in the EJB module.

**So the final question is:** If you agree with this EAR setup, than is it possible to provide custom ejb-jar.xml or glassfish-ejb-jar.xml or finally a persistence.xml descriptor ?

**Comment by lprimak** [comment]
I don't see application.xml in your example.  Perhaps you forgot to check this in?

Unfortunately, I still don't see what you are trying to achieve, at least not in a compliant way.
What are you trying to do exactly by overwriting ejb-jar.xml? 
in Modern Java EE, ejb-jar.xml should not be used at all.

**Comment by rsoika** [comment]
The application.xml is build by maven. 

Just as the "beans.xml" is used to influence the behavior of the CDI beans, so a ejb-jar. xml descriptor is needed to influence the behavior of EJBs. The deployment descriptor has the task of changing the annotations that the developer pretends.

Imaging you write EJB which is resizing a Image to a default size of 100x100 pixel. Put this EJB into a jar and make it public. With the help of a deployment descriptor a third party can influence the image size by providing other environment settings for this ejb without changing the source code of this ejb. That is what I am trying to do (in simplified terms). How would you solve this?
 
To sum it up: I do not think that it is a payara problem. I'm glad I can discuss this problem with you here at payara. But I believe that in Glassfish4 something went wrong with the classpath resolution in the EAR deployment.  The same thing can easily be archived in a war file.

**Comment by m-reza-rahman** [comment]
Ralph reached out to me and asked me to provide some nominal support on this issue. Ralph over the years has been a vocal supporter of the Java EE community, including sharing his Java EE adoption story as well as blogging. This appears to me to be a legitimate use case. If possible, can we fix this please? I know it's an area that is not well specified. Perhaps that can be fixed in Jakarta EE too but I don't think fixing it now would cause any Java EE 8 compatibility issues.

**Comment by lprimak** [comment]
Unfortunately, I either still don't understand the use case completely, or there are plenty of ways to accomplish this in a standardized way.
Extending ejb-jar in a way that is not what it was originally intended to do isn't the way to go, in my opinion.
My opinion that ejb-jar should be deprecated, not extended (over time of course)

Have you tried the following to accomplish your configuration goals?
- System properties
- Additional arguments to the API
- System property / Environment variable substitution in ejb-jar.xml
- Using JNDI configuration entries
- Using MicroProfile config

As far as merit, I took this issue on in the first place because I thought it had merit
However, upon spending a few hours with the example provided, I believe the methods above
are much more suited for the task, as opposed to doing something non-standard, and confusing.

I really tried hard to understand the use case, and if I cannot, I don't think many other people can either.

If there is indeed something simple here that I don't understand, please give me a concrete, simple case, in code perhaps.

**Comment by lprimak** [comment]
I still don't understand why my proposed solution isn't what you want.
I think I provided you with a complete working example of how you can do what
you desire.
I don't believe it's possible to do what you want exactly the way you want it.

This isn't standard and unsupported:
```
<JarModule>
	<groupId>com.rsoika</groupId>
	<artifactId>payara-issue-2508-lib</artifactId>
	<bundleDir>/</bundleDir>
</JarModule>
```

Also this:
```
<manifestEntries>
  <Class-Path>payara-issue-2508-lib-1.0.0.jar</Class-Path>
</manifestEntries>
```

I really think you are trying to shoehorn too many non-standard things together, and it won't work

**Comment by rsoika** [comment]
The sample contains the source code, of course. But imagine that you only have the 3rd party library with EJB components and you have to integrate it into one ear.  This is a serious use case. 

When you are in the role of a "Application Deployer" you need typically to customize deployment descriptors to overwrite annotated behavior. This is also the standard conform idea of Java EE - I believe. 
Just for example the 'glassfish-ejb-jar.xml' file. The component developer can't provide you with this file - ok? Simmilar is the persistence.xml which should also not be part of the 3rd party library. 

In a WAR we dont have any problem with this, because we have this central place to put all descriptors: /WEB-INF/. 
For that reason it's so easy to work with these components in a WAR. But in EAR we don't have this central configuration point. 

The idea with this manifest classpath trick was from the core developers from Glassfhish2 at sun microsystems - years ago. It was not my idea. But the idea is to pull-in EJB classes into a EJB module which is (as a Application Deployer) under my control. 

It would be great if another standard solution exits. Or maybe this is an issue for the future releases. 

I hope you can now understand the scenario: I am - as a component developer - providing you - as a application deployer - with generic EJB components. Its your freedom to overwrite my annotated behavior in your custom project. 

I really don't want to annoy you.  I just want to explain this use case, which is not unusual.

**Comment by lprimak** [comment]
Oh, no problem.
However, I still feel like you are not really listening to what I am saying.

To repeat:
Have you tried the following to accomplish your configuration goals?
- System properties
- Additional arguments to the API
- System property / Environment variable substitution in ejb-jar.xml
- Using JNDI configuration entries
- Using MicroProfile config

What I am trying to say that deployment descriptors **are not meant to be overwritten/be customization points** 
There are more modern ways to achieve configuration / customization and **deployment descriptors are not one of them**

If you feel that deployment descriptor is **the only way** to provide configuration, please provide a **concrete, simple** example of such a scenario after trying methods suggested above.

**Comment by lprimak** [comment]
Im going to close this issue.
If you need to discuss further and / or provide a concrete simple example how this can be easily remedied, please open another issue and reference this one.

**Comment by rsoika** [comment]
No, thanks, I think we exchanged everything just to see that we can't discuss the point together. As it is now closed it was in fact a waste of time.

**Comment by lprimak** [comment]
Sorry you feel that way. 
I am not able to take any concrete action with the information provided

**Comment by OndrejM** [comment]
@rsoika @m-reza-rahman thanks for reaching out. I went through the discussion and the example project and I think your usecase is valid but your solution is not standard and supported.

In short, the use case is that you want to package EJB components in a shared library and configure it later by a custom ejb-jar.xml if I understand correctly. 

The problem is that EJBs can only be placed in an EAR in an EJB-JAR or WAR modules and not in shared libraries. If you plase it in a shared library, it's included in all EJB-JAR and WAR modules - if you have more than one, the server tries to deploy it multiple times and fails.

Putting the EJB JAR into the root of EAR and hack the classpath may result in unexpected behavior which we won't fix because it's not supported.

I can recommend 2 solutions:

Solution 1:

If you can modify the JAR with EJBs, then make the EJB classes plain Java classes without EJB annotations and configure everything in ejb-jar.xml in an empty EJB JAR. This way you can add the shared code into the EAR as a shared lib and EJBs will be created only once.

Solution 2:

if you can't modify the JAR with EJBs, then the only thing you can do is turn it into an EJB JAR. Even if it doesn't contain ejb-jar.xml, you should be able to create a deployment plan that supplies new and overrides existing deployment descriptors, including ejb-jar.xml: https://docs.oracle.com/cd/E18930_01/html/821-2417/gilfm.html#gijyb

With this approach, you would configure deployment with an additional deployment plan JAR so you would end up with 2 deployment artifacts - an EAR and JAR.

Deployment plans are supported exactly for the purpose of overriding any configuration present in EARs and I think this solution suits best what you're seeking. If you fail with using a deployment plan, then please raise another issue and it would be a valid issue. But we can't accept your current approach as a valid usecase that needs fixing, I hope you understand.

**Comment by lprimak** [comment]
@OndrejM I forgot (and never used) deployment plan.  Perhaps there is a need for a blog.

**Comment by rsoika** [comment]
Thanks a lot for this clarification. The solution 2 is exactly what I have overlooked and this solution is of course acceptable. So we finally find successful an end for this thread. There is no bug and nothing is missing in Payara.

But some thoughts to JSR88: It seems that JSR88 is no longer part (or optional) in EE7. That's a pity. To me it seems that the idea of sharing EJB modules with others was never very popular in the enterprise environment. And maybe this is the reason why the interest in this concept got lost. Presumably, EARs are really out of date.  As I wrote, running the application in a single WAR works perfect and I can configure my components easily. 

So thinking more in the concepts of microservices as in deployment plans for monolithic EARs is the better way, I believe now. I will no longer follow the concept of EAR. 

@lprimak : yes there maybe a need for a blog, but this maybe only worthwhile, if you think that deployment plans and EARs are relevant.  You don't have to write this for me now ;-)
